### ISSUE TITLE
[QUESTION] Unexpected updates on unused token embeddings when using --untie-embeddings-and-output-weights

### ISSUE THREAD

**Issue by bugm** [NVIDIA/Megatron-LM#2994]
I have a question regarding embedding parameter updates during training, and I would like to confirm whether my understanding is correct or if there are other mechanisms involved.

**Setup**

- Vocabulary size: 128256
- During training, all tokens appearing in the dataset have token IDs strictly within [128000, 128256)
   Tokens with IDs < 128000 never appear in the input data
- Training is done with the flag:
  --untie-embeddings-and-output-weights

**My understanding**

Given the above setup, I would expect that:
For token IDs that never appear in any batch:

- Their corresponding rows in the embedding matrix should not receive gradients
- Therefore, their embedding parameters should remain unchanged throughout training

**Observed behavior**

However, I observe that:

- All rows of the embedding matrix change over the course of training, including those corresponding to token IDs that never appear in the data

- Moreover, these changes are not continuous per iteration
Instead, they seem to occur sporadically, e.g. once every hundreds or even thousands of iterations

I have double-checked the input data and can confirm that no token IDs outside [128000, 128256) are present

**Question**

I would like to understand what mechanisms could cause embedding weights of unused token IDs to change in this scenario. For example:

Could this be related to:

- Optimizer behavior (e.g. Adam moment updates, weight decay)?
- Is there any known behavior in Megatron-LM where embedding parameters may change even without direct gradient updates?

Any clarification or pointers to relevant code paths would be greatly appreciated. Thanks a lot for your help!

Best regards

**Comment by bugm** [comment]
Added information:
I add some custom code to record the word_embeddings.weight[15339,:20] (the first 20 dim of token "hello")

I  find that it first changes at iteariton 1258 and then it changes every dozens of iterations

<img width="1425" height="787" alt="Image" src="https://github.com/user-attachments/assets/2265f87a-488e-49fe-9807-a8169fc6652e" />

each time, one of the 20 dim will take a change at e-05 level and the are no input tokens with token id "15339"

<img width="1335" height="467" alt="Image" src="https://github.com/user-attachments/assets/7975d490-2dff-4b65-a531-65deb53591d3" />

<img width="1342" height="466" alt="Image" src="https://github.com/user-attachments/assets/b0c9e1d9-7fbd-4625-8076-6946eb11c6ab" />

<img width="1303" height="466" alt="Image" src="https://github.com/user-attachments/assets/d31e7fde-6aac-4afb-b22a-16c07e9fb179" />

finally, the word_embeddings.weight[15339,:20]  at iteration1

<img width="1072" height="116" alt="Image" src="https://github.com/user-attachments/assets/393b36ec-7d84-46ca-a6a2-a50a1931a467" />

at iteration  7267

<img width="1127" height="146" alt="Image" src="https://github.com/user-attachments/assets/dd454e8d-4dd8-498f-a57b-c83e6ecf9a8c" />

**Comment by maanug-nv** [comment]
Hi @bugm , thanks for your question. I can confirm that once the model is fully trained, we never see it generate token ids in the range [vocab size, vocab size + vocab padding). E.g. the output layer will be updated to produce very low scores for the padding tokens. Therefore, I think this is a non-issue.

**Comment by maanug-nv** [comment]
I just want to confirm that when you mention the 'embedding matrix/weights' you are referring to the input vocab embedding layer, correct?

**Comment by deepakn94** [comment]
Do you have weight decay turned on? If so, can you turn it off and see if embedding weights for the un-seen tokens are still changing?

**Comment by maanug-nv** [comment]
+1 on checking weight decay, if you can share a full reproducer script , that would also help to further investigate if needed
